src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs

```diff
@@ -2499,18 +2514,134 @@ public void Dispose()
                 return;
             }
 
-            this.DisposingEvent.SafeInvoke(this, EventArgs.Empty);
-            commandRuntime = null;
-            currentObjectInPipeline = null;
-            _input.Clear();
-            // _scriptBlock = null;
-            // _localsTuple = null;
-            // _functionContext = null;
+            ExecutionContext currentContext = LocalPipeline.GetExecutionContextFromTLS();
+            if (Context != currentContext && _scriptBlock.HasCleanupBlock)
```

Comment:
Same comment here about being able to save some cycles if `HasCleanupBlock` is false